File: internal/handler/audit.go
package handler

import (
	"encoding/json"

	"github.com/gin-gonic/gin"

	httpreq "github.com/zhaoxiaoyang741/HomeStock/internal/api/http/request"
	"github.com/zhaoxiaoyang741/HomeStock/internal/model"
	"github.com/zhaoxiaoyang741/HomeStock/internal/repository"
)

type changesPayload struct {
	Before any `json:"before,omitempty"`
	After  any `json:"after,omitempty"`
}

func marshalChanges(before, after any) string {
	b, err := json.Marshal(changesPayload{Before: before, After: after})
	if err != nil { return "" }
	return string(b)
}

func actorFromRequest(c *gin.Context) httpreq.Actor { return httpreq.From(c) }

func recordAuditLog(
	repo repository.AuditLogRepo,
	actor httpreq.Actor,
	action, entityType, entityID, entityName, changes string,
) {
	// Errors intentionally swallowed — audit failure must not break the primary operation.
	_ = repo.Create(&model.AuditLog{
		TenantID:      actor.TenantID,
		UserName:      actor.UserName,
		UserID:        actor.UserID,
		Channel:       actor.Channel,
		Action:        action,
		EntityType:    entityType,
		EntityID:      entityID,
		EntityName:    entityName,
		ChangesDetail: changes,
	})
}

File: web/src/types/audit.ts
export type AuditAction = 'create' | 'update' | 'delete'
export type AuditEntityType = 'item' | 'category'

export interface AuditLog {
  id: string
  tenant_id: string
  user_name: string
  user_id: string
  channel: string
  action: AuditAction
  entity_type: AuditEntityType
  entity_id: string
  entity_name: string
  changes_detail: string // JSON string: { before?: object, after?: object }
  created_at: string
}

export interface AuditLogListResponse {
  logs: AuditLog[]
  total: number
  page: number
  page_size: number
}

export interface AuditLogFilter {
  action?: string
  channel?: string
  user_name?: string
  start_date?: string
  end_date?: string
  page?: number
  page_size?: number
}

export interface ChangesDetail {
  before?: Record<string, unknown>
  after?: Record<string, unknown>
}

export function parseChangesDetail(raw: string): ChangesDetail {
  try {
    return JSON.parse(raw) as ChangesDetail
  } catch {
    return {}
  }
}
